### PR TITLE
blobstore: add the total bytes request when transfer logging is disabled

### PR DESCRIPTION
**Pull request by sandeepvinayak**
## Summary

Populate the directory-transfer response's `totalBytesTransferred` even when `transferStatusLoggingEnabled=false`, by accumulating per-object sizes during listing (downloads) or per-file size stat (uploads). Callers can now leave the per-file `TransferListener` off — which avoids major heap pressure on large directory operations — and still get a useful byte total on success.

Also lowers the per-file `transferComplete` / `onFinish` log line from INFO to DEBUG in both AWS (`S3LoggingTransferListener`) and Ali (`OssLoggingTransferListener`), so even when the listener is enabled it doesn't flood log appender queues on directories with thousands of files.

## Behavior

For both `DirectoryDownloadResponse.totalBytesTransferred` and `DirectoryUploadResponse.totalBytesTransferred`:

| `transferStatusLoggingEnabled` | Outcome | `totalBytesTransferred` |
|---|---|---|
| `true` | any | listener-accumulated total (unchanged) |
| `false` | full success (zero failed transfers) | sum of requested object/file sizes |
| `false` | partial failure (≥1 failed transfer) | `0` — callers consult `failedTransfers` |

## Changes

**AWS**
- `S3LoggingTransferListener` — per-file `transferComplete` log: INFO → DEBUG.
- `AwsTransformer.toDownloadDirectoryRequest` — collapsed into a single canonical 3-arg method; the existing `getPrefixExclusionsFilter` now optionally accumulates retained-object size into `totalBytesRequested` on the same true-return path, so excluded objects are never counted.
- `AwsTransformer.toUploadDirectoryRequest` — collapsed into a single canonical 3-arg method; the per-file `uploadFileRequestTransformer` now stats `Files.size(source)` into `totalBytesRequested` (best-effort; IOException is swallowed since the doomed file will surface via `failedTransfers`).
- `AwsAsyncBlobStore.doDownloadDirectory` / `doUploadDirectory` — allocate both counters and pick the response value via a shared `resolveDirectoryTotalBytes` helper.

**Ali**
- `OssLoggingTransferListener` — per-file `onFinish` log: INFO → DEBUG.
- `AliAsyncBlobStore.doDownloadDirectory` — accumulates `blob.getObjectSize()` in the listing consumer (after folder-marker + exclusion checks).
- `AliAsyncBlobStore.doUploadDirectory` — sums the existing stage-1 `fileSizes` stat pass.
- Shared `resolveDirectoryTotalBytes` helper mirrors the AWS one.

## Testing

- AWS: 4 new tests — listener-off success + partial-failure, for both download and upload (`mvn test -pl blob/blob-aws`: 258 passed, 2 pre-existing skips).
- Ali: 2 existing `...ReturnsNullTotal` tests renamed to `...ReturnsRequestedBytesOnSuccess` and re-asserted to expect the sum; 2 new partial-failure tests; OssLoggingTransferListener test updated for the DEBUG level on `onFinish` (`mvn test -pl blob/blob-ali`: 244 passed).

## Cross-cloud notes

- AWS S3 Transfer Manager: downloads exercise `DownloadDirectoryRequest.filter`; uploads exercise `uploadFileRequestTransformer` (no listing filter is exposed for uploads).
- Ali OSS: download already lists objects explicitly in `AliAsyncBlobStore` (so `getObjectSize()` is in hand during listing); upload already does a stage-1 stat pass — both reuse what's there.

## Compatibility

The `totalBytesTransferred` field shape is unchanged. Callers who previously saw `null` when the listener was off now see the requested total on success (or `0` on partial failure). The listener-on path is unchanged.

### CODE REVIEW

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/async/AwsAsyncBlobStore.java [hunk 1]
```diff
@@ -438,36 +438,64 @@ protected CompletableFuture<PresignedUrlResponse> doPresign(PresignedUrlRequest
...
+   */
+  private static Long resolveDirectoryTotalBytes(
+      boolean loggingEnabled,
+      boolean allTransfersSucceeded,
+      AtomicLong totalBytesTransferred,
+      AtomicLong totalBytesRequested) {
+    if (loggingEnabled) {
+      return totalBytesTransferred.get();
+    }
+    return allTransfersSucceeded ? totalBytesRequested.get() : 0L;
```
Comment (LihaoLiuXs) [inline blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/async/AwsAsyncBlobStore.java]: This changes the public contract: before this PR, `totalBytesTransferred` was `null` when logging was off — callers could use that to tell "I didn't ask for byte tracking." Now it's either a real number or `0L`, never `null`.

Any downstream code that does `if (response.getTotalBytesTransferred() == null)` to detect the "logging disabled" case will silently stop working. And `0L` on failure is ambiguous — it looks the same as an empty directory that succeeded.

Worth either keeping `null` for backward compat (and adding a separate `totalBytesRequested` field), or bumping the version and calling this out as a breaking change in the release notes.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java [hunk 2]
```diff
@@ -669,47 +680,64 @@ public PresignedUrlResponse toPresignedUrlResponse(
...
     }
 
-    // If we have prefixes to exclude from the download, then add in a filter here
-    if (request.getPrefixesToExclude() != null && !request.getPrefixesToExclude().isEmpty()) {
-      downloadDirectoryRequestBuilder.filter(
-          getPrefixExclusionsFilter(request.getPrefixesToExclude()));
-    }
-    return downloadDirectoryRequestBuilder.build();
-  }
+    builder.filter(getPrefixExclusionsFilter(
```
Comment (LihaoLiuXs) [inline blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java]: Heads up — the old code only called `builder.filter(...)` when there were actual prefixes to exclude. Now it always installs a filter (even if it's just a pass-through that returns `true` for everything).

S3 Transfer Manager might treat "no filter set at all" differently from "a filter that happens to allow everything" (e.g. short-circuiting internal logic, or the filter lambda's overhead on very large listings). Might be safer to only set the filter when `prefixesToExclude` is non-empty OR `totalBytesRequested` is non-null, and skip it entirely otherwise.

**Review by LihaoLiuXs — COMMENTED** [review]
Reviewed two areas — see inline comments.

**Review by kchoy-sfdc — COMMENTED** [review]
Three findings from review — 1 major (dormant footgun), 2 minor. Posting as `COMMENT`, not a block — none of these rise to a change-request bar.

- **M1 (major, dormant today):** `transferStatusLoggingEnabled` is guarded asymmetrically between `toUploadDirectoryRequest` and `toDownloadDirectoryRequest`. Today's only call site always passes a non-null counter, so the upload-side `&& totalBytesTransferred != null` is silent — but a future caller passing `null` to opt out of byte counting would silently lose the listener with no error. Suggested fix: pick one rule and apply symmetrically.
- **m2 (minor):** Filter side effect on `totalBytesRequested` is not idempotent. Today's SDK calls the filter once per listed object so it's correct, but a retry inside listing pagination could double-count. Either document the SDK assumption or de-dupe.
- **m3 (minor):** Swallowed `IOException` from `Files.size` will silently under-report `totalBytesRequested` if the assumption "every IOException corresponds to a file S3TM will also fail" doesn't hold for a transient I/O hiccup. Suggested fix: log at WARN/DEBUG, or document that the count may under-report.

Nothing here blocks the PR — the major is a footgun for future maintenance, not a present bug. Detailed inline comments below.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java [hunk 3]
```diff
@@ -741,60 +780,33 @@ public UploadDirectoryRequest toUploadDirectoryRequest(DirectoryUploadRequest re
 
     boolean hasTags = request.getTags() != null && !request.getTags().isEmpty();
     boolean hasObjectLock = request.getObjectLock() != null;
+    boolean transferStatusLoggingEnabled =
+        request.isTransferStatusLoggingEnabled() && totalBytesTransferred != null;
```
Comment (kchoy-sfdc) [inline blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java:798]: **[MAJOR]** `transferStatusLoggingEnabled` is checked with inconsistent rules on the upload vs. download paths — the listener flag is silently dropped on upload if a future caller happens to pass a `null` `totalBytesTransferred`.

Here on the upload path:

```java
boolean transferStatusLoggingEnabled =
    request.isTransferStatusLoggingEnabled() && totalBytesTransferred != null;
```

But `toDownloadDirectoryRequest` (around `:711`) does not AND the request flag with `totalBytesTransferred != null`:

```java
if (request.isTransferStatusLoggingEnabled()) {
  S3LoggingTransferListener transferListener =
      S3LoggingTransferListener.create(totalBytesTransferred);
  ...
}
```

The two methods now treat the same flag differently. Today the only production call site (`AwsAsyncBlobStore.doUploadDirectory`) always supplies a non-null counter, so the inconsistency is dormant — but it is exactly the kind of footgun that bites a future caller (or a test) that constructs a transformer call with `null` to opt out of byte counting and then loses the listener as a side effect, with no error or log.

Pick one rule and apply it in both directions. Either of these would resolve it:

1. Drop the `&& totalBytesTransferred != null` guard so a logging request always builds the listener. The listener constructor takes an `AtomicLong` directly and would NPE early on `null`.
2. Keep the guard but apply the symmetric one in `toDownloadDirectoryRequest` so the contract holds in both directions.

Whichever option you pick, also add a unit test that calls `toUploadDirectoryRequest(request, null, requested)` with `isTransferStatusLoggingEnabled() == true` and asserts the resulting behavior — there is no test pinning that path today, so the inconsistency would survive a refactor.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java [hunk 4]
```diff
@@ -669,47 +680,64 @@ public PresignedUrlResponse toPresignedUrlResponse(
...
+    List<String> excludes = prefixesToExclude != null ? prefixesToExclude : List.of();
     return s3Object -> {
-      for (String prefixToExclude : prefixesToExclude) {
+      for (String prefixToExclude : excludes) {
         if (s3Object.key().startsWith(prefixToExclude)) {
           return false;
         }
       }
+      if (totalBytesRequested != null) {
+        totalBytesRequested.addAndGet(s3Object.size());
```
Comment (kchoy-sfdc) [inline blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java:753]: **[MINOR]** The filter is invoked once per listed object today, but its side effect on `totalBytesRequested` is not idempotent.

```java
return s3Object -> {
  for (String prefixToExclude : excludes) {
    if (s3Object.key().startsWith(prefixToExclude)) {
      return false;
    }
  }
  if (totalBytesRequested != null) {
    totalBytesRequested.addAndGet(s3Object.size());
  }
  return true;
};
```

S3 Transfer Manager's documented contract for `DownloadDirectoryRequest.filter` is to call it exactly once per listed object during the listing phase — so under normal operation the count is exactly the included objects' size. However: (a) the contract is not formally idempotent-safe (a future SDK release that retries the listing internally could double-call the filter), and (b) if the listing pagination itself retries an entire page, the filter may see the same object twice. For the listener-off success path this matters because `totalBytesRequested` becomes the response's `totalBytesTransferred` directly.

Two options:

1. **Document the SDK assumption.** Add a one-line comment to the filter lambda noting that callers rely on "filter invoked once per retained object" semantics and that retries within a page would over-count. Cheap; matches current SDK reality.
2. **De-duplicate by key.** Track keys already counted in a `Set<String>` and skip the `addAndGet` on a repeat. Adds heap that this PR is explicitly trying to avoid — a regression in spirit — so only do this if you find evidence the SDK can re-invoke the filter.

I'd lean toward option 1 unless you find evidence of the SDK re-calling the filter under retries. Worth confirming the same about the upload `uploadFileRequestTransformer` — the SDK does not document whether it is invoked once per file or once per file *attempt*, and the upload-side `Files.size` accumulation has the same shape of risk.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java [hunk 5]
```diff
@@ -804,11 +816,26 @@ public UploadDirectoryRequest toUploadDirectoryRequest(
...
+          if (countRequested) {
+            // Best-effort stat, evaluated independently per file. A file that throws here is
+            // also a file the SDK can't upload, so it'll appear in failedTransfers and the
+            // response will already report 0 via resolveDirectoryTotalBytes — the under-counted
+            // requested total is never the value handed back to the caller in that case.
+            Path source = fileRequestBuilder.build().source();
+            if (source != null) {
+              try {
+                totalBytesRequested.addAndGet(Files.size(source));
+              } catch (IOException ignored) {
```
Comment (kchoy-sfdc) [inline blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java:848]: **[MINOR]** The swallowed `IOException` from `Files.size` will silently under-report `totalBytesRequested` if the assumption "every `IOException` here corresponds to a file that S3TM will also fail to upload" ever doesn't hold.

```java
if (countRequested) {
  Path source = fileRequestBuilder.build().source();
  if (source != null) {
    try {
      totalBytesRequested.addAndGet(Files.size(source));
    } catch (IOException ignored) {
      // Skip this file's contribution; ...
    }
  }
}
```

The inline justification (a file that throws on `Files.size` will fail the upload too and surface in `failedTransfers`, at which point `resolveDirectoryTotalBytes` already returns `0`) is sound for the **happy-path full-success** case — but it relies on a strong invariant: *every* `IOException` from `Files.size` corresponds to a file that S3TM will also fail to upload. That is plausible but not guaranteed; transient I/O errors are real, and S3TM may retry past them. If exactly that mismatch occurs (size stat fails, upload succeeds), the response under-counts by that file with no signal to the operator.

Two cheap improvements (one is enough):

1. Log the `IOException` at WARN (or at least DEBUG) so an operator can correlate an unexpected discrepancy. Right now the swallow is completely silent.
2. Note in the existing inline comment that the count may under-report in this case, so the next reader doesn't assume it is always exact.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/async/AwsAsyncBlobStore.java (hunk 1 above)
Comment (sandeepvinayak) [reply blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/async/AwsAsyncBlobStore.java]: `null` make more sense given the method returns the Object Long but it's fine to put in bytes if known - the customers who do not depend on it doesn't loss anything,and the one depending on it will still get it.

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java (hunk 4 above)
Comment (sandeepvinayak) [reply blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java:753]: it's fallback cheap algorithm when the client opts out from listener and is not expected to be accurate for others reasons as well such as file IOException in case of file uploads

File: blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java (hunk 3 above)
Comment (sandeepvinayak) [reply blob/blob-aws/src/main/java/com/salesforce/multicloudj/blob/aws/AwsTransformer.java:798]: good catch!

**Review by LihaoLiuXs — APPROVED** [review]
